src/main/scala/dev/kamu/cli/commands/PullCommand.scala

```diff
@@ -15,20 +16,35 @@ class PullCommand(
   metadataRepository: MetadataRepository,
   sparkRunner: SparkRunner,
   datasetIDs: Seq[String],
-  all: Boolean
+  all: Boolean,
+  recursive: Boolean
 ) extends Command {
   private val logger = LogManager.getLogger(getClass.getName)
 
   def run(): Unit = {
-    val datasets =
-      if (all)
-        metadataRepository.getAllDatasets()
-      else
-        datasetIDs
-          .map(DatasetID)
-          .map(metadataRepository.getDataset)
-
-    val numUpdated = datasets
+    val dependsOn: DatasetID => List[DatasetID] = {
+      if (all || recursive) {
+        metadataRepository.getDataset(_).dependsOn.toList
+      } else {
+        metadataRepository
+          .getDataset(_)
+          .dependsOn
+          .toList
+          .filter(datasetIDs.map(DatasetID).contains)
+      }
+    }
+
+    val plan = {
+      val depGraph = new DependencyGraph[DatasetID](dependsOn)
+      depGraph.resolve(datasetIDs.map(DatasetID).toList)
+    }
+
+    logger.debug(s"Pulling datasets in following order:")
+    plan.foreach(d => logger.debug(s"  ${d.toString}"))
+
+    // Use Spark to pull each item in the plan
+    val numUpdated = plan
+      .map(metadataRepository.getDataset)
```

Comment:
Ugh, kinda sucks that we'll be loading them the second time, perhaps if per my other suggestion we isolate the whole resolve step into a method we could cache the `Dataset`s as `dependsOn` loads the definitions.

Reply:
Not all of them are loaded 2 times. Not sure if dependsOn needs cache. If it does I think it should be somewhere in MetadataRepository.